File: fs/src/main/scala/farjs/fs/FSFoldersHistory.scala
package farjs.fs

import farjs.filelist.history._
import scommons.react._
import scommons.react.hooks._

import scala.concurrent.ExecutionContext.Implicits.global
import scala.scalajs.js

case class FSFoldersHistoryProps(currDirPath: String)

object FSFoldersHistory extends FunctionComponent[FSFoldersHistoryProps] {

  val foldersHistoryKind: HistoryKind = HistoryKind("farjs.folders", 100)

  protected def render(compProps: Props): ReactElement = {
    val historyProvider = HistoryProvider.useHistoryProvider
    val props = compProps.wrapped
    val currDirPath = props.currDirPath

    useLayoutEffect({ () =>
      if (currDirPath.nonEmpty) {
        for {
          foldersHistory <- historyProvider.get(foldersHistoryKind).toFuture
          _ <- foldersHistory.save(History(currDirPath, js.undefined)).toFuture
        } yield ()
      }
      ()
    }, List(currDirPath))

    null
  }
}

File: fs/FSFoldersHistory.mjs
/**
 * @import { HistoryKind } from "@farjs/filelist/history/HistoryProvider.mjs"
 */
import { useLayoutEffect } from "react";
import HistoryProvider from "@farjs/filelist/history/HistoryProvider.mjs";

/**
 * @typedef {{
 *  readonly currDirPath: string;
 * }} FSFoldersHistoryProps
 */

/**
 * @param {FSFoldersHistoryProps} props
 */
const FSFoldersHistory = (props) => {
  const { foldersHistoryKind } = FSFoldersHistory;

  const historyProvider = HistoryProvider.useHistoryProvider();
  const currDirPath = props.currDirPath;

  const saveHistory = async () => {
    const foldersHistory = await historyProvider.get(foldersHistoryKind);
    await foldersHistory.save({ item: currDirPath });
  };

  useLayoutEffect(() => {
    if (currDirPath.length > 0) {
      saveHistory();
    }
  }, [currDirPath]);

  return null;
};

FSFoldersHistory.displayName = "FSFoldersHistory";

/** @type {HistoryKind} */
FSFoldersHistory.foldersHistoryKind = {
  name: "farjs.folders",
  maxItemsCount: 100,
};

export default FSFoldersHistory;
